SO-3882: Template model and API

nagyo opened the pull request.

nagyo committed:
SO-3882: Add ability to define additional component deletion policies...

... in implementations of TerminologyRepositoryConfigurers

https://snowowl.atlassian.net/browse/SO-3882

nagyo committed:
SO-3882: Fix license header

https://snowowl.atlassian.net/browse/SO-3882

nagyo committed:
SO-3882: Fix invalid type in deletion policy rule

https://snowowl.atlassian.net/browse/SO-3882

nagyo committed:
SO-3889: Move Charmatcher for invalid file name characters to FileUtils

https://snowowl.atlassian.net/browse/SO-3889

nagyo committed:
Merge remote-tracking branch 'origin/7.x' into feature/SO-3882_template_model

nagyo committed:
Merge remote-tracking branch 'origin/7.x' into feature/SO-3882_template_model

nagyo committed:
Merge remote-tracking branch 'origin/7.x' into feature/SO-3882_template_model

nagyo committed:
[snomed] Fix permission level of SNOMED component create requests

nagyo committed:
Merge remote-tracking branch 'origin/7.x' into feature/SO-3882_template_model

nagyo committed:
[request] Remove unnecessary method override

nagyo committed:
SO-3885: Allow SnomedBulkRequest instantiation outside of package

https://snowowl.atlassian.net/browse/SO-3885

nagyo committed:
Merge remote-tracking branch 'origin/7.x' into feature/SO-3882_template_model

nagyo committed:
[commons] Extend constructor to find resources relative to class ...

... in case the test bundle is a fragment

nagyo committed:
Merge remote-tracking branch 'origin/7.x' into feature/SO-3882_template_model

nagyo committed:
[datastore] Clean up CompositeComponentDeletionPolicy

apeteri commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/repository/CompositeComponentDeletionPolicy.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class CompositeComponentDeletionPolicy implements ComponentDeletionPolicy {
+
+	Map<Class<?>, Predicate<RevisionDocument>> deletionPolicies;
```
Probably doesn't make too much of a difference, but `Predicate<RevisionDocument>` occurrences could be replaced with `ComponentDeletionPolicy` to really drive the point home of this class being a _composite_ of other policies.

apeteri commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/repository/CompositeComponentDeletionPolicy.java:
```diff
@@ -0,0 +1,68 @@
+	@Override
+	public boolean canDelete(final RevisionDocument revision) {
+		return deletionPolicies.entrySet().stream()
+				.filter(entry -> entry.getKey().isInstance(revision))
```
If there is no "is the instance class same or subclass of this class"-kind of testing needed (where you are registering multiple applicable policies, eg. for `SnomedConceptDocument` → `SnomedComponentDocument` → `SnomedDocument`), lookup instead of iteration would be quicker:

```java
if (revision == null) { return false; }
return deletionPolicies.get(revision.getClass()).test(revision);
```

nagyo replied:
There is a need for those checks though, the main SNOMED deletion policy is applicable for all subtypes of SnomedDocument. I will add a fast evaluation part to the beginning of that method, that could make processing a bit faster.

apeteri commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/repository/RepositoryBuilder.java:
```diff
@@ -145,6 +145,20 @@ public Repository build(Environment env) {
+			
+			repositoryConfigurers.forEach(configurer -> {
+				configurer.getComponentDeletionPolicies().forEach( (clazz, predicate) -> {
+					if (!compositePolicy.getDeletionPolicies().containsKey(clazz)) {
```
`compositePolicy.getDeletionPolicies().putIfAbsent(clazz, predicate);`

It's also a bit unexpected (for me) to see a builder on `CompositeComponentDeletionPolicy`, suggesting that the result is immutable, then proceed to pull a writable `Map` out via the get method and inject extra entries. A factory method could be added to the builder that starts with an existing `CompositeComponentDeletionPolicy`.

apeteri commented on core/com.b2international.snowowl.datastore/src/com/b2international/snowowl/core/repository/TerminologyRepositoryConfigurer.java:
```diff
@@ -32,4 +36,9 @@
+	/**
+	 * @return additional component deletion policies for subtypes of RevisionDocument
+	 */
+	Map<Class<?>, Predicate<RevisionDocument>> getComponentDeletionPolicies();
```
This can also be hidden entirely behind the `ComponentDeletionPolicy` interface:

```suggestion
	ComponentDeletionPolicy getComponentDeletionPolicy();
```

nagyo replied:
But we need that class information to be able to select the appropriate policy for a given type, don't we?

apeteri replied:
My mistake; wanted to say `CompositeComponentDeletionPolicy`. This ties into my previous comment, this could be done if there would be a way to merge multiple composites into one, eg. via a `CompositeComponentDeletionPolicy#builder(CompositeComponentDeletionPolicy)` or `CompositeComponentDeletionPolicy.Builder#withPolicies(CompositeComponentDeletionPolicy)` method.

nagyo committed:
SO-3882: Refactor CompositeComponentDeletionPolicy

- use ComponentDeletionPolicy instead of Predicate
- add mergeWith method to combine multiple composite policies

https://snowowl.atlassian.net/browse/SO-3882
